### ISSUE TITLE
Sinkhorn-Distance in Machine Learning (Based on pytorch)

### ISSUE THREAD

**Issue by ChrisNieo** [rflamary/POT#119]
As we know that both Tensorflow and Pytorch implement back propagation by calculating gradient through the computational flow graph. We must use torch.Tensor in every step of calculation so Pytorch can draw the computational flow graph and get the gradient. But how can I use this distance  (calculated by using numpy and including C++) in Pytorch during back propagation?

**Comment by rflamary** [comment]
Hello,

We do not provide yet pytorch/tensorflow implementation of our OT solvers. We paln on doing that in the future but do not yet have some code commited or in PR.

you can find a pytorch implementation of sinkhorn here
https://github.com/rythei/PyTorchOT

and a nn.Module imlplementation here
https://gist.github.com/wohlert/8589045ab544082560cc5f8915cc90bd

**Comment by kilianFatras** [comment]
Hello,

I also suggest to have a look at the really good geomloss library : https://github.com/jeanfeydy/geomloss.

Wasserstein distance is not implemented in the library but you can use the Sinkhorn divergence with a small regularization parameter to have an idea of the results.

**Comment by ChrisNieo** [comment]
> Hello,
> 
> We do not provide yet pytorch/tensorflow implementation of our OT solvers. We paln on doing that in the future but do not yet have some code commited or in PR.
> 
> you can find a pytorch implementation of sinkhorn here
> https://github.com/rythei/PyTorchOT
> 
> and a nn.Module imlplementation here
> https://gist.github.com/wohlert/8589045ab544082560cc5f8915cc90bd

Thank you very much! I'll see and have a try. :)

**Comment by rflamary** [comment]
I strongly agree with @kilianFatras ,

Geomloss is very efficient and scale very well compared to the simple implementation I recommended earlier. I can't believe I forgot about it ;).

**Comment by ChrisNieo** [comment]
Ok! I'll try. If i meet any questions, i'm afraid i will bother you again.
